### PR TITLE
Refactoring org.apache.cloudstack.network.tungsten.service

### PR DESCRIPTION
**Pull request by gzhao9**
### Description
This PR addresses the issue mentioned in #8087 by refactoring the test suites in three classes: `TungstenElementTest.java`, `TungstenFabricUtilsTest.java`, and `TungstenServiceImplTest.java`. The issue involves redundant mock object creations for `TungstenAnswer` in these test suites.

#### Proposed Changes
I propose refactoring the test code to eliminate redundancy and improve maintainability. Specifically, I suggest creating a reusable `CreateMockTungstenAnswer` class for creating `TungstenAnswer` mock objects with consistent stub behavior. 

#### Benefits of the Proposed Refactoring:

- **As a result, we can reduce 321 lines of code, which makes the test cases smaller, easier to understand, and better to maintain.**
- **After refactoring, if the `TungstenAnswer` mock creation needs to be modified in the future, the developer only needs to modify one time in `CreateMockTungstenAnswer` instead of 151 places of duplicate mock creation in 3 testclasses.**

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// under the License.
+package org.apache.cloudstack.network.tungsten.service;
+
+import org.apache.cloudstack.network.tungsten.agent.api.TungstenAnswer;
+import org.mockito.Mock;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class CreateMockTungstenAnswer {
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java:25]: would a more appropriate name be `MockTungstenAnswerFactory`?

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    TungstenAnswer tungstenAnswer;
+    CreateMockTungstenAnswer(boolean Result){
+        tungstenAnswer=mock(TungstenAnswer.class);
+        when(tungstenAnswer.getResult()).thenReturn(Result);
+    }
+    CreateMockTungstenAnswer(){
+        when(tungstenAnswer.getResult()).thenReturn(true);
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java:33]:
```suggestion
        this(true);
```

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    TungstenAnswer get(){return tungstenAnswer;}
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java:35]:
```suggestion
    TungstenAnswer get() {
        return tungstenAnswer;
    }

```

**Review by DaanHoogland — COMMENTED** [review]
good change, some build error though. I added some suggestions as well.

**Comment by DaanHoogland** [comment]
not sure if the build errors are related to this PR

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java (hunk 1 above)
Comment (gzhao9) [reply plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/CreateMockTungstenAnswer.java:25]: Good suggestion, I've renamed `CreateMockTungstenAnswer` to `MockTungstenAnswerFactory` and recommitted.

**Comment by gzhao9** [comment]
> not sure if the build errors are related to this PR

I found the error reason; the parameter used `Result`, which doesn't follow camel case naming. I've made an update.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
> > not sure if the build errors are related to this PR
> 
> I found the error reason; the parameter used `Result`, which doesn't follow camel case naming. I've made an update.

it still doesn't compile @gzhao9 , can you have a look?

**Comment by gzhao9** [comment]
> > > not sure if the build errors are related to this PR
> > 
> > 
> > I found the error reason; the parameter used `Result`, which doesn't follow camel case naming. I've made an update.
> 
> it still doesn't compile @gzhao9 , can you have a look?

This is because the most recent update involved a file that I had committed.

I've made the update and made sure it works on my end.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@gzhao9 have a look at https://github.com/apache/cloudstack/actions/runs/6560826849/job/17824151994?pr=8098#step:6:9444 it seems you have some line endings mixed (\n instead of \r?)

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by gzhao9** [comment]
> according to https://github.com/apache/cloudstack/actions/runs/6564274866/job/17850770581?pr=8098#step:6:9439 you still have mixed line endings @gzhao9 .

I'll keep trying to deal with it. BTW I was very confused. Another PR I submitted did not have this formatting problem.
Also, I'm curious how you guys generally deal with this issue on Windows.

**Comment by DaanHoogland** [comment]
> > according to https://github.com/apache/cloudstack/actions/runs/6564274866/job/17850770581?pr=8098#step:6:9439 you still have mixed line endings @gzhao9 .
> 
> I'll keep trying to deal with it. BTW I was very confused. Another PR I submitted did not have this formatting problem. Also, I'm curious how you guys generally deal with this issue on Windows.

I don´t work on windows recently, but I remember a tool called d2u or dos2unix. it can do the conversions to and from. nowadays there is also the use of a VM that could help ;)

**Comment by gzhao9** [comment]
@DaanHoogland To avoid also having the ending problem. I submitted two more PRs #8137 and #8139. also with similar idea to solve the duplicate creation problem of mock. If you are interested you can take a look.

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenFabricUtilsTest.java [hunk 4]
```diff
@@ -93,13 +92,12 @@ public void sendTungstenCommandWithNullAnswer() {
...
         TungstenCommand tungstenCommand = mock(TungstenCommand.class);
 
         when(tungstenProviderDao.findByZoneId(anyLong())).thenReturn(tungstenProviderVO);
         when(agentMgr.easySend(anyLong(), any(TungstenCommand.class))).thenReturn(tungstenAnswer);
-        when(tungstenAnswer.getResult()).thenReturn(false);
 
         tungstenFabricUtils.sendTungstenCommand(tungstenCommand, anyLong());
     }
-}
+}
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenFabricUtilsTest.java:103]:
```suggestion
}
```

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenServiceImplTest.java [hunk 5]
```diff
@@ -1759,11 +1662,10 @@ public void removeNetworkGatewayFromLogicalRouterTest() {
...
-        TungstenAnswer tungstenAnswer = mock(TungstenAnswer.class);
+        TungstenAnswer tungstenAnswer = new MockTungstenAnswerFactory(true).get();
 
         when(tungstenFabricUtils.sendTungstenCommand(any(DeleteTungstenRoutingLogicalRouterCommand.class), anyLong())).thenReturn(tungstenAnswer);
-        when(tungstenAnswer.getResult()).thenReturn(true);
 
         assertTrue(tungstenService.deleteLogicalRouter(1L, "948f421c-edde-4518-a391-09299cc25dc2"));
     }
-}
+}
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenServiceImplTest.java:1671]:
```suggestion
}
```

File: plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenElementTest.java [hunk 6]
```diff
@@ -1039,4 +1011,4 @@ public void commitMigration() {
         tungstenElement.commitMigration(nic, network, vm, src, dest);
         verify(tungstenFabricUtils, times(1)).sendTungstenCommand(any(DeleteTungstenVRouterPortCommand.class), anyLong());
     }
-}
+}
```
Comment (DaanHoogland) [inline plugins/network-elements/tungsten/src/test/java/org/apache/cloudstack/network/tungsten/service/TungstenElementTest.java:1014]:
```suggestion
}
```

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
unit tests only , GHA should suffice.

**Review by rohityadavcloud — APPROVED** [review]

**Pull request merged**